### ISSUE TITLE
Kokkos on Volta Numerics Issue (reported by Stan Moore)

### ISSUE THREAD

**Issue by stanmoore1** [kokkos/kokkos#1455]
I'm seeing another Volta Kokkos numerics issue in ExaMiniMD using the latest development branch at 6b2a50b61bc6afd666cdd6dda527fd0d3369a70b. When I build for `KOKKOS_ARCH=Volta70` on a Titan V, RHEL7, x86_64, I get non-deterministic behavior. If I instead compile with `KOKKOS_ARCH=Pascal60` and run on the same GPU, the non-deterministic behavior goes away. Also, if I change the team size from 20 to 32 while using `KOKKOS_ARCH=Volta70`, the non-deterministic behavior also goes away.

Details: 

ExaMiniMD is a proxy app for LAMMPS, see https://github.com/ECP-copa/ExaMiniMD
Using the SNAP potential
To compile: `make -j32 KOKKOS_ARCH=SNB,Volta70 KOKKOS_DEVICES=Cuda CXX=~/kokkos/bin/nvcc_wrapper MPI=0`
To run `cd input/snap; ../../src/ExaMiniMD -il in.snap.W`

Correct output using `KOKKOS_ARCH=Pascal60` or teamsize=32:
```
#Timestep Temperature PotE ETot Time Atomsteps/s
0 300.000000 0.000000 0.038764 0.000000 0.000000e+00
1 291.331234 0.000000 0.037644 0.126724 -9.150532e+00
2 267.039481 0.000000 0.034505 0.251948 2.191273e+04
3 231.905150 0.000000 0.029965 0.376618 2.201011e+04
4 192.766560 0.000000 0.024908 0.501345 2.200005e+04
5 157.089220 0.000000 0.020298 0.626065 2.200128e+04
6 131.425697 0.000000 0.016982 0.750701 2.201611e+04
7 120.104839 0.000000 0.015519 0.875066 2.206409e+04
8 124.436182 0.000000 0.016079 0.999349 2.207864e+04
9 142.595576 0.000000 0.018425 1.123660 2.207367e+04
10 170.199592 0.000000 0.021992 1.248367 2.200358e+04
```

Example of non-deterministic behavior using `KOKKOS_ARCH=Volta70` and teamsize=20
```
#Timestep Temperature PotE ETot Time Atomsteps/s
0 300.000000 0.000000 0.038764 0.000000 0.000000e+00
1 293.459085 0.000000 0.037919 0.152231 1.802524e+04
2 275.096847 0.000000 0.035546 0.301987 1.832314e+04
3 246.552142 0.000000 0.031858 0.451673 1.833183e+04
4 214.150839 0.000000 0.027671 0.601072 1.836680e+04
5 184.784356 0.000000 0.023877 0.750229 1.839672e+04
6 164.265392 0.000000 0.021225 0.899448 1.838908e+04
7 158.157140 0.000000 0.020436 1.048356 1.842749e+04
8 167.678052 0.000000 0.021666 1.197640 1.838107e+04
9 190.479308 0.000000 0.024612 1.346993 1.837258e+04
10 223.385892 0.000000 0.028864 1.496283 1.838033e+04
```

Second example of non-deterministic behavior:
```
#Timestep Temperature PotE ETot Time Atomsteps/s
0 300.000000 0.000000 0.038764 0.000000 0.000000e+00
1 293.496886 0.000000 0.037924 0.150637 1.821598e+04
2 273.891298 0.000000 0.035390 0.300329 1.833097e+04
3 244.494617 0.000000 0.031592 0.449866 1.834997e+04
4 211.140295 0.000000 0.027282 0.599423 1.834752e+04
5 181.343515 0.000000 0.023432 0.748653 1.838772e+04
6 161.456289 0.000000 0.020862 0.897761 1.840277e+04
7 155.712809 0.000000 0.020120 1.047076 1.837726e+04
8 164.693241 0.000000 0.021281 1.196360 1.838107e+04
9 186.541164 0.000000 0.024104 1.345941 1.834470e+04
10 218.770568 0.000000 0.028268 1.495371 1.836299e+04
```

Modified input script:
```patch
diff --git a/input/snap/in.snap.W b/input/snap/in.snap.W
index b19472d..5acbd79 100644
--- a/input/snap/in.snap.W
+++ b/input/snap/in.snap.W
@@ -7,8 +7,8 @@ units           metal
 newton                 on
 # generate the box and atom positions using a BCC lattice
 
-lattice         sc 3.1803
-region         box block 0 4 0 8 0 8
+lattice         sc 2.3
+region          box block 0 14 0 14 0 14
 create_box     1 box
 create_atoms   1 box
 
@@ -22,7 +22,7 @@ pair_coeff * * W.snapcoeff W W.snapparam W
 
 # Setup output
 
-thermo         10
+thermo         1
 
 # Set up NVE run
 
@@ -33,6 +33,6 @@ neigh_modify once no every 1 delay 0 check yes
 
 velocity all create 300.0 4928459 loop geom
 fix 1 all nve
-run            100
+run            10
 ```

**Comment by crtrott** [comment]
Looks like already in step 0 the forces are wrong ...

**Comment by crtrott** [comment]
Found it: the single statements need an appropriate (sub)warp sync. 
Maybe we should just snapshot ExaMiniMD into Kokkos. The amount of bugs this SNAP potential exposes is kinda amazing ...

**Comment by crtrott** [comment]
I put a pull request up and spot check is running

**Comment by crtrott** [comment]
Ah great: this actually fixed Clang6 with CUDA 9.0 build using clang as the CUDA compiler for Volta. Now all the tests pass for it.

**Comment by stanmoore1** [comment]
@crtrott This fixed the issue I was seeing in LAMMPS SNAP as well. Looks like you got it.
